darshan-runtime/lib/darshan-heatmap.c

```diff
@@ -0,0 +1,398 @@
+#define HEATMAP_UNLOCK() pthread_mutex_unlock(&heatmap_runtime_mutex)
+#endif
+
+/* note that if the break condition is triggered in this macro, then it
+ * will exit the do/while loop holding a lock that will be released in
+ * POST_RECORD().  Otherwise it will release the lock here (if held) and
+ * return immediately without reaching the POST_RECORD() macro.
+ */
+/* NOTE: unlike other modules, the PRE_RECORD here does not attempt to
+ * initialize this module if it isn't already.  That should have been done
+ * in the _register() call before reaching this point.  Skipping the
+ * initialization attempt here makes it safe to use atomics or spinlocks
+ * in the critical wrapper path.
+ */
+#define HEATMAP_PRE_RECORD() do { \
+    HEATMAP_LOCK(); \
+    if(heatmap_runtime && !heatmap_runtime->frozen) break; \
+    HEATMAP_UNLOCK(); \
+    return(ret); \
+} while(0)
+
+/* same as above but for void fns */
+#define HEATMAP_PRE_RECORD_VOID() do { \
+    HEATMAP_LOCK(); \
+    if(heatmap_runtime && !heatmap_runtime->frozen) break; \
+    HEATMAP_UNLOCK(); \
+    return; \
+} while(0)
+
+#define HEATMAP_POST_RECORD() do { \
+    HEATMAP_UNLOCK(); \
+} while(0)
+
+static void heatmap_output(
+    void **heatmap_buf,
+    int *heatmap_buf_sz)
+{
+    struct darshan_heatmap_record* rec;
+    void* contig_buf_ptr;
+    int i;
+    double end_timestamp;
+    unsigned long this_size;
+    int tmp_nbins;
+
+    HEATMAP_LOCK();
+    assert(heatmap_runtime);
+
+    *heatmap_buf_sz = 0;
+
+    heatmap_runtime->frozen = 1;
+    end_timestamp = darshan_core_wtime();
+
+    contig_buf_ptr = *heatmap_buf;
+
+    /* iterate through records (heatmap histograms) */
+    for(i=0; i<heatmap_runtime->rec_count; i++)
+    {
+        rec = (struct darshan_heatmap_record*)((uintptr_t)*heatmap_buf + i*(sizeof(*rec) + DARSHAN_MAX_HEATMAP_BINS*2*sizeof(int64_t)));
+
+        tmp_nbins= ceil(end_timestamp/rec->bin_width_seconds);
+
+        /* are there bins beyond the execution time of the program? */
+        if(tmp_nbins < rec->nbins)
+        {
+            /* truncate bins so that we don't report any beyond the time when
+             * instrumentation stopped
+             */
+            rec->nbins = tmp_nbins;
+            /* shift read_bins down so that memory remains contiguous even
+             * if nbins has been reduced
+             */
+            memmove(&rec->write_bins[rec->nbins], rec->read_bins,
+                rec->nbins*sizeof(int64_t));
+            rec->read_bins = &rec->write_bins[rec->nbins];
+        }
+
+        /* now shift the entire record + bins as a contiguous block down in
+         * the buffer so that the entire buffer is contiguous
+         */
+        this_size = sizeof(*rec) + rec->nbins * 2 * sizeof(uint64_t);
+        memmove(contig_buf_ptr, rec, this_size);
+        contig_buf_ptr += this_size;
+        *heatmap_buf_sz += this_size;
+    }
+
+    HEATMAP_UNLOCK();
+
+    return;
+}
+
+static void heatmap_cleanup()
+{
```

Comment:
That's interesting.  I got tripped up here by being over-reliant on address sanitizer, which apparently doesn't complain because the pointers are all technically still reachable at exit.  I switched to valgrind `valgrind --leak-check=full --show-leak-kinds=all` and now I see the leaks there.  Will fix.

Reply:
Fixed in ccefac9d67703efe29da8e6fd05e93482dbd1e0e